main/src/mpu.cpp

```diff
@@ -0,0 +1,157 @@
+
+    // Calibration on start?
+    // mpudev.CalibrateAccel(10);
+    // mpudev.CalibrateGyro(10);
+
+    // Alternatively, we can set offsets obtained by running IMU_Zero
+    // example from the library.
+    // mpudev.setXAccelOffset(0);
+    // mpudev.setYAccelOffset(0);
+    // mpudev.setZAccelOffset(0);
+    // mpudev.setXGyroOffset(0);
+    // mpudev.setYGyroOffset(0);
+    // mpudev.setZGyroOffset(0);
+
+    if (!mpudev.testConnection()) {
+        // die
+        Serial.println("Connection to mpu failed");
+        return;
+    }
+
+    if (mutex == nullptr) {
+        // die
+        Serial.println("Couldn't allocate resources for mutex!");
+        return;
+    }
+
+    init_success = true;
+    accel_res = mpudev.get_acce_resolution() / 2;
+    gyro_res = mpudev.get_gyro_resolution() / 2;
+}
+
+/**
+ * @brief Consume and return collected data.
+ *        ! CAN BLOCK INDEFINITELY (if INCLUDE_vTaskSuspend == 1) !
+ * @return Max g force, Average g force, Average rotation.
+ */
+Data get_mpu() {
+    xSemaphoreTake(mutex, portMAX_DELAY);
+
+    Data dane{
+        .max_x = internal.max_x,
+        .max_y = internal.max_y,
+        .max_z = internal.max_z,
+        .avg_x = internal.avg_x / count, // Instead of counting could implement
+        .avg_y = internal.avg_y / count, // moving average, but dunno if it's
+        .avg_z = internal.avg_z / count, // superior and/or viable alternative.
+        .rot_x = internal.rot_x / count,
+        .rot_y = internal.rot_y / count,
+        .rot_z = internal.rot_z / count};
+
+    count = 0;
+    internal = {};
+
+    xSemaphoreGive(mutex);
+
+    return dane;
+}
+
+/**
+ * @brief Main MPU task that collects data from the sensor.
+ *        Die gracefully if sensor not initialised.
+ *        ! CAN BLOCK INDEFINITELY (if INCLUDE_vTaskSuspend == 1) !
+ */
+void mpu_task([[maybe_unused]] void *pvParameters) {
+    int16_t ax, ay, az, gx, gy, gz;
+    float fax, fay, faz, fgx, fgy, fgz;
+
+    if (!init_success)
+        vTaskDelete(NULL);
+
+    for (;;) {
+        mpudev.getMotion6(&ax, &ay, &az, &gx, &gy, &gz);
+
+        fax = abs(ax * accel_res);
+        fay = abs(ay * accel_res);
+        faz = abs(az * accel_res);
+        fgx = abs(gx * gyro_res);
+        fgy = abs(gy * gyro_res);
+        fgz = abs(gz * gyro_res);
+
+        xSemaphoreTake(mutex, portMAX_DELAY);
+
+        if (internal.max_x < fax)
+            internal.max_x = fax;
+        if (internal.max_y < fay)
+            internal.max_y = fay;
+        if (internal.max_z < faz)
+            internal.max_z = faz;
+
+        internal.avg_x += fax;
+        internal.avg_y += fay;
+        internal.avg_z += faz;
+        internal.rot_x += fgx;
+        internal.rot_y += fgy;
+        internal.rot_z += fgz;
+
+        count++;
+
+        xSemaphoreGive(mutex);
+
+        vTaskDelay(pdMS_TO_TICKS(100));
+    }
+}
+
+/**
+ * @brief Consume and display collected data.
+ */

```

Comment:
remove, also it doesn't consume any data as it's working on  a copy of the internal Data buffer